src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java

```diff
@@ -226,22 +228,45 @@ public void claimTask(String taskId,
      */
     @Transactional
     public void unclaimTask(String taskId, AccessControlResponse accessControlResponse) {
+        final boolean granularPermissionFeatureEnabled = launchDarklyFeatureFlagProvider
+            .getBooleanValue(
+                FeatureFlag.GRANULAR_PERMISSION_FEATURE,
+                accessControlResponse.getUserInfo().getUid(),
+                accessControlResponse.getUserInfo().getEmail()
+            );
+
         String userId = accessControlResponse.getUserInfo().getUid();
-        PermissionRequirements permissionsRequired = PermissionRequirementBuilder.builder().buildSingleType(MANAGE);
+        PermissionRequirements permissionsRequired;
+        if (granularPermissionFeatureEnabled) {
+            permissionsRequired = PermissionRequirementBuilder.builder()
+                .buildSingleRequirementWithOr(UNCLAIM, UNASSIGN);
+        } else {
+            permissionsRequired = PermissionRequirementBuilder.builder().buildSingleType(MANAGE);
+        }
+
         boolean taskHasUnassigned;
         final boolean isFeatureEnabled = launchDarklyFeatureFlagProvider
             .getBooleanValue(
                 FeatureFlag.RELEASE_2_ENDPOINTS_FEATURE,
                 accessControlResponse.getUserInfo().getUid(),
                 accessControlResponse.getUserInfo().getEmail()
             );
+
         if (isFeatureEnabled) {
             TaskResource taskResource = roleAssignmentVerification.verifyRoleAssignments(
                 taskId, accessControlResponse.getRoleAssignments(), permissionsRequired
             );
             String taskState = taskResource.getState().getValue();
             taskHasUnassigned = taskState.equals(CFTTaskState.UNASSIGNED.getValue());
 
+            if (granularPermissionFeatureEnabled && !userId.equals(taskResource.getAssignee())) {
+                PermissionRequirements unassignPermissionsRequired
+                    = PermissionRequirementBuilder.builder().buildSingleType(UNASSIGN);
+                roleAssignmentVerification.verifyRoleAssignments(
```

Comment:
Can we check if unassign flag is set to true in TaskRoleResource object that returned by previous verifyRoleAssignments call, instead of calling it again?